src/dictionary.rs

```diff
@@ -0,0 +1,445 @@
+/// use std::collections::HashMap;
+///
+/// use chewing::{dictionary::Dictionary, syl, zhuyin::Bopomofo};
+///
+/// let mut dict = HashMap::new();
+/// let dict_mut = dict.as_mut_dict().unwrap();
+/// dict_mut.insert(&[syl![Bopomofo::C, Bopomofo::E, Bopomofo::TONE4]], ("測", 100).into())?;
+/// # Ok(())
+/// # }
+/// ```
+pub trait DictionaryMut {
+    fn insert(
+        &mut self,
+        syllables: &[Syllable],
+        phrase: Phrase<'static>,
+    ) -> Result<(), DictionaryUpdateError>;
+
+    fn update(
+        &mut self,
+        syllables: &[Syllable],
+        phrase: Phrase,
+        user_freq: u32,
+        time: u64,
+    ) -> Result<(), DictionaryUpdateError>;
+
+    fn remove(
+        &mut self,
+        syllables: &[Syllable],
+        phrase_str: &str,
+    ) -> Result<(), DictionaryUpdateError>;
+}
+
+#[derive(Error, Debug)]
+#[error("build dictionary error")]
+pub struct BuildDictionaryError {
+    #[from]
+    source: Box<dyn std::error::Error + Send + Sync>,
+}
+
+impl From<std::io::Error> for BuildDictionaryError {
+    fn from(source: std::io::Error) -> Self {
+        BuildDictionaryError {
+            source: Box::new(source),
+        }
+    }
+}
+
+pub trait DictionaryBuilder {
+    fn set_info(&mut self, info: DictionaryInfo) -> Result<(), BuildDictionaryError>;
+    fn insert(
+        &mut self,
+        syllables: &[Syllable],
+        phrase: Phrase,
+    ) -> Result<(), BuildDictionaryError>;
+    fn build(&mut self, path: &Path) -> Result<(), BuildDictionaryError>;
+}
+
+impl Dictionary for HashMap<Vec<Syllable>, Vec<Phrase<'static>>> {
+    fn lookup_phrase(&self, syllables: &[Syllable]) -> Phrases {
+        self.get(syllables)
+            .cloned()
+            .map(|v| Box::new(v.into_iter()) as Phrases)
+            .unwrap_or_else(|| Box::new(std::iter::empty()))
+    }
+
+    fn entries(&self) -> DictEntries {
+        Box::new(
+            self.iter()
+                .flat_map(|(k, v)| v.iter().map(|phrase| (k.clone(), phrase.clone()))),
+        )
+    }
+
+    fn about(&self) -> DictionaryInfo {
+        Default::default()
+    }
+
+    fn as_mut_dict(&mut self) -> Option<&mut dyn DictionaryMut> {
+        Some(self)
+    }
+}
+
+impl DictionaryMut for HashMap<Vec<Syllable>, Vec<Phrase<'static>>> {
+    fn insert(
+        &mut self,
+        syllables: &[Syllable],
+        phrase: Phrase<'static>,
+    ) -> Result<(), DictionaryUpdateError> {
+        let vec = self.entry(syllables.to_vec()).or_default();
+        if vec.iter().any(|it| it.as_str() == phrase.as_str()) {
+            return Err(DictionaryUpdateError {
+                source: Box::new(DuplicatePhraseError),
+            });
+        }
+        vec.push(phrase);
+        Ok(())
+    }
+
+    fn update(
+        &mut self,
+        _syllables: &[Syllable],
+        _phrase: Phrase,
+        _user_freq: u32,
+        _time: u64,
+    ) -> Result<(), DictionaryUpdateError> {
+        Ok(())
```

Comment:
This should either document it will be ignored or just return an error.

Reply:
The implementation was missing 😓 
The HashMap implementation serves as an model and can be used for other doc comments. I'll revisit this when writing the doc for the `DictionaryMut` trait.